.github/workflows/ci.yml

```diff
@@ -24,20 +24,114 @@ on:
 
 jobs:
   check-semantic-version:
-    if: github.event.pull_request.draft == false
+    if: github.ref != 'refs/heads/main'
     uses: octue/workflows/.github/workflows/check-semantic-version.yml@main
     with:
       path: pyproject.toml
       breaking_change_indicated_by: minor
 
-  run-tests:
+  check-main-merge:
+    if: github.ref != 'refs/heads/main'
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Check branch is ahead of main
+        run: |
+          if ! git merge-base --is-ancestor origin/main ${{ github.event.pull_request.head.sha }};
+          then echo "::error::This branch is not up-to-date with the latest main branch commit.";
+          exit 1; fi
+
+  lint:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+
+      - name: Setup python
+        uses: actions/setup-python@v5
+        with:
+          python-version: '3.11'
+
+      - name: Install and configure poetry
+        uses: snok/install-poetry@v1
+        with:
+          virtualenvs-create: true
+          virtualenvs-in-project: true
+
+      - name: Setup virtual environment cache
+        id: cached-poetry-dependencies
+        uses: actions/cache@v4
+        with:
+          path: .venv
+          key: venv-${{ runner.os }}-3.11-${{ hashFiles('**/poetry.lock') }}
+
+      - name: Install dependencies (if not cached)
+        if: steps.cached-poetry-dependencies.outputs.cache-hit != 'true'
+        run: poetry install --only dev --no-interaction --no-root
+
+      - name: Install root project
+        run: poetry install --no-interaction
+
+      - name: Run precommit
+        run: SKIP=build-docs,check-branch-name poetry run pre-commit run --all-files
+
+  publish-test:
```

Comment:
With skip-existing, it still checks that the wheel is buildable and uploadable with `twine` - ie checks the things that you actually could mess up - so it's still worth running on subsequent executions of the workflow.

So overall I disagree, because if this is the thing that's failing for whatever reason, you want to be able to iterate quite quickly on that without waiting for an entire test matrix each time.

Also (as in this case) decoupling it means I can actually test that it'll publish, without the tests passing - I may actually make this release with failing tests then follow on with a test fix release. Whilst not ideal, I think it's better to keep projects progressing than hold them up because a test that was previously failing still fails, an this allows that.